File: app/dto/conversation.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Dict, Any, List, Optional
from enum import Enum

class MessageType(Enum):
    USER_MESSAGE = "USER_MESSAGE"
    MANAGER_MESSAGE = "MANAGER_MESSAGE"
    WORKFLOW_UPDATE = "WORKFLOW_UPDATE"
    WORKER_UPDATE = "WORKER_UPDATE"
    APPROVAL_REQUEST = "APPROVAL_REQUEST"
    USER_INPUT_REQUEST = "USER_INPUT_REQUEST"
    DECISION_REQUEST = "DECISION_REQUEST"
    ERROR = "ERROR"
    COMPLETION = "COMPLETION"

@dataclass
class ConversationMessage:
    id: str
    conversation_id: str
    message_type: MessageType
    content: str
    sender: str
    timestamp: str
    metadata: Dict[str, Any] = field(default_factory=dict)
    related_event_id: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "conversation_id": self.conversation_id,
            "message_type": self.message_type.value,
            "content": self.content,
            "sender": self.sender,
            "timestamp": self.timestamp,
            "metadata": self.metadata,
            "related_event_id": self.related_event_id
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> ConversationMessage:
        return cls(
            id=data["id"],
            conversation_id=data["conversation_id"],
            message_type=MessageType(data["message_type"]),
            content=data["content"],
            sender=data["sender"],
            timestamp=data["timestamp"],
            metadata=data.get("metadata", {}),
            related_event_id=data.get("related_event_id")
        )

@dataclass
class Conversation:
    id: str
    project_id: str
    title: str
    messages: List[ConversationMessage]
    created_at: str
    updated_at: str
    is_active: bool

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "project_id": self.project_id,
            "title": self.title,
            "messages": [m.to_dict() for m in self.messages],
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "is_active": self.is_active
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> Conversation:
        messages = [ConversationMessage.from_dict(m) for m in data.get("messages", [])]
        return cls(
            id=data["id"],
            project_id=data["project_id"],
            title=data["title"],
            messages=messages,
            created_at=data["created_at"],
            updated_at=data["updated_at"],
            is_active=data.get("is_active", True)
        )

File: client/lib/models/conversation.dart
enum MessageType {
  userMessage('USER_MESSAGE'),
  managerMessage('MANAGER_MESSAGE'),
  workflowUpdate('WORKFLOW_UPDATE'),
  workerUpdate('WORKER_UPDATE'),
  approvalRequest('APPROVAL_REQUEST'),
  userInputRequest('USER_INPUT_REQUEST'),
  decisionRequest('DECISION_REQUEST'),
  error('ERROR'),
  completion('COMPLETION');

  final String value;
  const MessageType(this.value);

  static MessageType fromString(String val) {
    return MessageType.values.firstWhere(
      (e) => e.value == val,
      orElse: () => MessageType.managerMessage,
    );
  }
}

class ChatMessage {
  final String id;
  final String conversationId;
  final MessageType messageType;
  final String content;
  final String sender;
  final String timestamp;
  final Map<String, dynamic> metadata;
  final String? relatedEventId;

  const ChatMessage({
    required this.id,
    required this.conversationId,
    required this.messageType,
    required this.content,
    required this.sender,
    required this.timestamp,
    this.metadata = const {},
    this.relatedEventId,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    return ChatMessage(
      id: json['id'] as String? ?? '',
      conversationId: json['conversation_id'] as String? ?? '',
      messageType: MessageType.fromString(json['message_type'] as String? ?? 'MANAGER_MESSAGE'),
      content: json['content'] as String? ?? '',
      sender: json['sender'] as String? ?? 'Manager',
      timestamp: json['timestamp'] as String? ?? '',
      metadata: (json['metadata'] as Map<String, dynamic>?) ?? const {},
      relatedEventId: json['related_event_id'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'conversation_id': conversationId,
    'message_type': messageType.value,
    'content': content,
    'sender': sender,
    'timestamp': timestamp,
    'metadata': metadata,
    'related_event_id': relatedEventId,
  };
}

class ChatConversation {
  final String id;
  final String projectId;
  final String title;
  final List<ChatMessage> messages;
  final String createdAt;
  final String updatedAt;
  final bool isActive;

  const ChatConversation({
    required this.id,
    required this.projectId,
    required this.title,
    this.messages = const [],
    required this.createdAt,
    required this.updatedAt,
    this.isActive = true,
  });

  factory ChatConversation.fromJson(Map<String, dynamic> json) {
    final rawMsgs = json['messages'] as List<dynamic>? ?? [];
    return ChatConversation(
      id: json['id'] as String? ?? '',
      projectId: json['project_id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      messages: rawMsgs.map((m) => ChatMessage.fromJson(m as Map<String, dynamic>)).toList(),
      createdAt: json['created_at'] as String? ?? '',
      updatedAt: json['updated_at'] as String? ?? '',
      isActive: json['is_active'] as bool? ?? true,
    );
  }

  ChatConversation copyWith({
    String? id,
    String? projectId,
    String? title,
    List<ChatMessage>? messages,
    String? createdAt,
    String? updatedAt,
    bool? isActive,
  }) {
    return ChatConversation(
      id: id ?? this.id,
      projectId: projectId ?? this.projectId,
      title: title ?? this.title,
      messages: messages ?? this.messages,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isActive: isActive ?? this.isActive,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'project_id': projectId,
    'title': title,
    'messages': messages.map((m) => m.toJson()).toList(),
    'created_at': createdAt,
    'updated_at': updatedAt,
    'is_active': isActive,
  };
}
